txn/buffer.go

```diff
@@ -0,0 +1,358 @@
+// For a given transaction, it must not be called until after a final oplog
+// entry has been passed to AddOp and it must not be called more than once.
+func (b *Buffer) GetTxnStream(m Meta) (<-chan db.Oplog, <-chan error) {
+	b.Lock()
+	defer b.Unlock()
+
+	opChan := make(chan db.Oplog)
+	errChan := make(chan error, 1)
+
+	if b.stopped {
+		return sendErrAndClose(opChan, errChan, ErrBufferClosed)
+	}
+
+	if !m.IsTxn() {
+		return sendErrAndClose(opChan, errChan, ErrNotTransaction)
+	}
+
+	state := b.txns[m.id]
+	if state == nil {
+		return sendErrAndClose(opChan, errChan, fmt.Errorf("GetTxnStream found no state for %v", m.id))
+	}
+
+	// The final oplog entry must have been passed to AddOp before calling this
+	// method, so we know this will be able to make progress.
+	<-state.ingestDone
+
+	if state.ingestErr != nil {
+		return sendErrAndClose(opChan, errChan, state.ingestErr)
+	}
+
+	// Launch streaming goroutine
+	b.wg.Add(1)
+	state.wg.Add(1)
+	go b.streamer(state, opChan, errChan)
+
+	return opChan, errChan
+}
+
+func (b *Buffer) streamer(state *txnState, opChan chan<- db.Oplog, errChan chan<- error) {
+LOOP:
+	for _, op := range state.buffer {
+		select {
+		case opChan <- op:
+		case <-state.stopChan:
+			errChan <- ErrTxnAborted
+			break LOOP
+		}
+	}
+	close(opChan)
+	close(errChan)
+	state.wg.Done()
+	b.wg.Done()
+}
+
+// PurgeTxn closes any transaction streams in progress and deletes all oplog
+// entries associated with a transaction.
+//
+// Must not be called concurrently with other transaction-related operations.
+// For a given transaction, it must not be called until after a final oplog
+// entry has been passed to AddOp and it must not be called more than once.
+func (b *Buffer) PurgeTxn(m Meta) error {
+	b.Lock()
+	defer b.Unlock()
+	if b.stopped {
+		return ErrBufferClosed
+	}
+	state := b.txns[m.id]
+	if state == nil {
+		return fmt.Errorf("PurgeTxn found no state for %v", m.id)
+	}
+
+	// When the lock is dropped, we don't want Stop to find this transaction and
+	// double-close it.
+	delete(b.txns, m.id)
+	close(state.stopChan)
+
+	// Wait for goroutines to terminate, then clean up.
+	state.wg.Wait()
+	state.purge()
+
+	return nil
+}
+
+// Stop shuts down processing and cleans up.  Subsequent calls to Stop() will return nil.
+// All other methods error after this is called.
+func (b *Buffer) Stop() error {
+	b.Lock()
+	if b.stopped {
+		b.Unlock()
+		return nil
+	}
+
+	b.stopped = true
+	for _, state := range b.txns {
+		close(state.stopChan)
+	}
+
+	b.Unlock()
+
+	// At this point we know any subsequent public method will see the buffer
+	// is stopped, no new goroutines will be launched, and existing goroutines
+	// have been signaled to close.  Next, wait for goroutines to stop, then
+	// clean up.
+
+	b.wg.Wait()
+	errs := make([]error, 0)
+	for _, state := range b.txns {
+		err := state.purge()
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	if len(errs) != 0 {
+		return errs[0]
```

Comment:
Do we need to track all errors in a potentially large `errs` slice if we're only ever returning the first one?